Corona-Warn-App/src/main/java/de/rki/coronawarnapp/diagnosiskeys/download/DownloadDiagnosisKeysTask.kt

```diff
@@ -0,0 +1,235 @@
+
+            return object : Task.Result {}
+        } catch (error: Exception) {
+            Timber.tag(TAG).e(error)
+
+            rollback(rollbackItems)
+
+            throw error
+        } finally {
+            Timber.i("Finished (isCanceled=$isCanceled).")
+            internalProgress.close()
+        }
+    }
+
+    private fun saveTimestamp(
+        currentDate: Date,
+        rollbackItems: MutableList<RollbackItem>
+    ) {
+        val lastFetchDateForRollback = LocalData.lastTimeDiagnosisKeysFromServerFetch()
+        rollbackItems.add {
+            LocalData.lastTimeDiagnosisKeysFromServerFetch(lastFetchDateForRollback)
+        }
+        Timber.tag(TAG).d("dateUpdate(currentDate=%s)", currentDate)
+        LocalData.lastTimeDiagnosisKeysFromServerFetch(currentDate)
+    }
+
+    private fun retrieveToken(rollbackItems: MutableList<RollbackItem>): String {
+        val googleAPITokenForRollback = LocalData.googleApiToken()
+        rollbackItems.add {
+            LocalData.googleApiToken(googleAPITokenForRollback)
+        }
+        return UUID.randomUUID().toString().also {
+            LocalData.googleApiToken(it)
+        }
+    }
+
+    private fun noKeysFetchedToday(): Boolean {
+        val currentDate = DateTime(timeStamper.nowUTC, DateTimeZone.UTC)
+        val lastFetch = DateTime(
+            LocalData.lastTimeDiagnosisKeysFromServerFetch(),
+            DateTimeZone.UTC
+        )
+        return (LocalData.lastTimeDiagnosisKeysFromServerFetch() == null ||
+            currentDate.withTimeAtStartOfDay() != lastFetch.withTimeAtStartOfDay()).also {
+            if (it) {
+                Timber.tag(TAG)
+                    .d("No keys fetched today yet (last=%s, now=%s)", lastFetch, currentDate)
+                BackgroundWorkHelper.sendDebugNotification(
+                    "Start Task",
+                    "No keys fetched today yet \n${DateTime.now()}\nUTC: $currentDate"
+                )
+            }
+        }
+    }
+
+    private fun rollback(rollbackItems: MutableList<RollbackItem>) {
+        try {
+            Timber.tag(TAG).d("Initiate Rollback")
+            for (rollbackItem: RollbackItem in rollbackItems) rollbackItem.invoke()
+        } catch (rollbackException: Exception) {
+            Timber.tag(TAG).e(rollbackException, "Rollback failed.")
+        }
+    }
+
+    private suspend fun getAvailableKeyFiles(requestedCountries: List<String>?): List<File> {
+        val availableKeyFiles =
+            keyFileDownloader.asyncFetchKeyFiles(if (environmentSetup.useEuropeKeyPackageFiles) {
+                listOf("EUR")
+            } else {
+                requestedCountries
+                    ?: appConfigProvider.getAppConfig().supportedCountries
+            }.map { LocationCode(it) })
+
+        if (availableKeyFiles.isEmpty()) {
+            Timber.tag(TAG).w("No keyfiles were available!")
+        }
+        return availableKeyFiles
+    }
+
+    private fun checkCancel() {
```

Comment:
Why is this function added in all Tasks? If implementations differ from task to task it could be at least added to the interface.
What do you think @d4rken ?

Reply:
Think it's okay like this, the task interface should be simple to allow for maximum flexibility. 
We can think about creating an `abstract class SmartTask : Task` but that's also not something that needs to be done here. 
Moving the `checkCancel` into a common base class is also not without issues.